lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp

```diff
@@ -701,6 +702,8 @@ LoopPipelinerInternal::emitEpilogue(RewriterBase &rewriter,
     for (Operation *op : opOrder) {
       if (stages[op] < i)
         continue;
+      if (llvm::dyn_cast<triton::gpu::InstructionBlockSchedOp>(op))
+        continue;
```

Comment:
passby: the expander shouldn't have special treatment for ops

Reply:
Hi @ThomasRaoux, 

Agree. The problem is that the scheduling op needs to be applied only in the body of the main loop. If we remove these lines then the expander will copy the sched. instructions to the prologue and epilogue of the gemm kernel which is not supposed to happen.

Do you have some ideas what we can do here?

Reply:
do you need to insert the op before pipelining?

Reply:
> do you need to insert the op before pipelining?

Hmm.. The alternative would be to create a pass which is going to inspect all `scf.for` Ops and check whether they contain `tt.dot` Ops. @ThomasRaoux, @antiagainst, what do you think?

Reply:
Why that's the case? You can still walk the IR and anchor on `tt.dot` like what you do right now? You can check whether there is an enclosing loop for sure--that's just a simple `getParentOfType` call.

Reply:
@antiagainst, thanks. Done